pkg/hub/addon/discovery_controller.go

```diff
@@ -0,0 +1,213 @@
+	recorder events.Recorder,
+) factory.Controller {
+	c := &addOnFeatureDiscoveryController{
+		clusterClient: clusterClient,
+		clusterLister: clusterInformer.Lister(),
+		addOnLister:   addOnInformers.Lister(),
+		recorder:      recorder,
+	}
+
+	return factory.New().
+		WithInformers(clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			clusterInformer.Informer()).
+		WithInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				key, _ := cache.MetaNamespaceKeyFunc(obj)
+				return key
+			},
+			addOnInformers.Informer()).
+		WithSync(c.sync).
+		ToController("AddOnFeatureDiscoveryController", recorder)
+}
+
+func (c *addOnFeatureDiscoveryController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	queueKey := syncCtx.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(queueKey)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return nil
+	}
+
+	if len(namespace) > 0 {
+		// sync a particular addon
+		return c.syncAddOn(ctx, namespace, name)
+	}
+
+	// sync all labels of the managed cluster
+	clusterName := name
+	cluster, err := c.clusterLister.Get(clusterName)
+	if errors.IsNotFound(err) {
+		// cluster is deleted
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Do not update addon label if cluster is deleting
+	if !cluster.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	newLabels := map[string]string{}
+	for key, value := range cluster.Labels {
+		if strings.HasPrefix(key, addOnFeaturePrefix) {
+			continue
+		}
+		newLabels[key] = value
+	}
+	addOns, err := c.addOnLister.ManagedClusterAddOns(clusterName).List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("unable to list addOns of cluster %q: %w", clusterName, err)
+	}
+	for _, addOn := range addOns {
+		// addon is deleting
+		if !addOn.DeletionTimestamp.IsZero() {
+			continue
+		}
+		key := addOnFeaturePrefix + addOn.Name
+		newLabels[key] = getAddOnLabelValue(addOn)
+	}
+
+	// no work if the cluster labels have no change
+	if (len(newLabels) == 0 && len(cluster.Labels) == 0) || reflect.DeepEqual(newLabels, cluster.Labels) {
+		return nil
+	}
+
+	// otherwise, update cluster
+	newCluster := cluster.DeepCopy()
+	newCluster.Labels = newLabels
+	_, err = c.clusterClient.ClusterV1().ManagedClusters().Update(ctx, newCluster, metav1.UpdateOptions{})
+	return err
+}
+
+func (c *addOnFeatureDiscoveryController) syncAddOn(ctx context.Context, clusterName, addOnName string) error {
+	klog.V(4).Infof("Reconciling addOn %q", addOnName)
+
+	addOn, err := c.addOnLister.ManagedClusterAddOns(clusterName).Get(addOnName)
+	if errors.IsNotFound(err) {
+		// addon is deleted
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
+	}
+	if err != nil {
+		return err
+	}
+
+	// addon is deleting
+	if !addOn.DeletionTimestamp.IsZero() {
+		return c.deleteAddOnLabel(ctx, clusterName, addOnName)
```

Comment:
it seems to me that using https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourcemerge/object_merger.go#L138 can be much easier.

Reply:
Code updated.